### PR TITLE
website: Upgrades to Docusaurus 3

### PR DESCRIPTION
**Pull request by vikram-dagger**
This commit upgrades the docs website to Docusaurus 3.0 using a clean install.

Key changes:
- Creates a new directory structure for docs as `docs/current_docs`, where `docs` is the Docusarus project root (contains the files from previous `website` dir) and `current_docs` is the Markdown content (contains the Markdown files from previous `docs/current` dir). 
- Removes `v0.1` and `v0.2` directories from the build and from all sidebar navigation. Retains only `current` and `zenith` versions in the build, representing the current and next version of Dagger.
- Uses Docusaurus built-in versioning system for `current` version (in `docs/current_docs`) and `zenith` version (in `versioned_docs/version-zenith`)
- Removes Docusaurus `includes` plugin in favour of native Docusaurus markdown import mechanism.
- Removes Docusaurus `dotenv` plugin as it does not seem to be required.
- Flattens directory structure for Docusaurus custom components as there are now fewer components and therefore additional categorization using sub-directories seems unnecessary (moves components from `src/components/[atoms,molecules]/*` to `src/components` and CSS files from `src/css/[atoms,molecules]/*` to `src/css`).
- Renames `.md` files to `.mdx` for all docs in `current` and `zenith` versions
- Removes all images and content related to the CUE SDK 
- Removes the theme files in `src/theme` and uses the default Docusaurus theme.
- Adds redirections for older docs (`/0.1` and `/0.2` URL paths) to `archive.docs.dagger.io`
- Removes `BrowserOnly` component, see https://github.com/dagger/dagger/pull/6164/commits/0bb93f09c41faefb2ddefe468a11e4df6b08078e)

Fixes DEV-2848

### CODE REVIEW

**Comment by vikram-dagger** [comment]
I'm running into this error, any ideas @slumbering ?

```
Error: Unable to build website for locale en.
    at tryToBuildLocale 
  [cause]: Error: Failed to compile due to Webpack errors.
  Error: Docusaurus server-side rendering could not render static page with path /docs/1202/plan because of error: Cannot destructure property 'daggerVersionLatestRelease' of '(0 , _docusaurus_useGlobalData__WEBPACK_IMPORTED_MODULE_1__.eZ)(...)' as it is undefined.
  
  --------------------------
  
  Error: Docusaurus server-side rendering could not render static page with path /docs/cli/389936/run-pipelines-cli because of error: Expected component `BrowserOnly` to be defined: you likely forgot to import, pass, or provide it.
  
  --------------------------
  
  Error: Docusaurus server-side rendering could not render static page with path /docs/cli/465058/install because of error: Expected component `BrowserOnly` to be defined: you likely forgot to import, pass, or provide it.
  
  --------------------------
  
  Error: Docusaurus server-side rendering could not render static page with path /docs/quickstart/729236/cli because of error: Expected component `BrowserOnly` to be defined: you likely forgot to import, pass, or provide it.
  
  --------------------------
  
  Error: Docusaurus server-side rendering could not render static page with path /docs/sdk/cue/347598/plan because of error: Cannot destructure property 'daggerVersionLatestRelease' of '(0 , _docusaurus_useGlobalData__WEBPACK_IMPORTED_MODULE_1__.eZ)(...)' as it is undefined.
  
  --------------------------
  
  Error: Docusaurus server-side rendering could not render static page with path /docs/sdk/cue/470907/get-started because of error: window is not defined
  It looks like you are using code that should run on the client-side only.
  To get around it, try using `<BrowserOnly>` (https://docusaurus.io/docs/docusaurus-core/#browseronly) or `ExecutionEnvironment` (https://docusaurus.io/docs/docusaurus-core/#executionenvironment).
  It might also require to wrap your client code in `useEffect` hook and/or import a third-party library dynamically (if any).
  
  --------------------------
  
  Error: Docusaurus server-side rendering could not render static page with path /docs/sdk/cue/526369/install because of error: window is not defined
  It looks like you are using code that should run on the client-side only.
  To get around it, try using `<BrowserOnly>` (https://docusaurus.io/docs/docusaurus-core/#browseronly) or `ExecutionEnvironment` (https://docusaurus.io/docs/docusaurus-core/#executionenvironment).
  It might also require to wrap your client code in `useEffect` hook and/or import a third-party library dynamically (if any).
  
  --------------------------
  
  Error: Docusaurus server-side rendering could not render static page with path /docs/zenith/user/install because of error: Expected component `BrowserOnly` to be defined: you likely forgot to import, pass, or provide it.


[INFO] Docusaurus version: 3.0.0
Node version: v18.16.0
```

**Comment by vikram-dagger** [comment]
I [removed the BrowserOnly component](https://github.com/dagger/dagger/pull/6164/commits/0bb93f09c41faefb2ddefe468a11e4df6b08078e) and [switched the implementation to JS](https://github.com/dagger/dagger/pull/6164/commits/941ed838092c9ef742447c3188108c3344846050) instead of TS to see if this would make a difference. After doing this, I see a similar error as before, but related to the `TabItem` component this time.

```
 [cause]: Error: Failed to compile due to Webpack errors.
  Error: Docusaurus server-side rendering could not render static page with path /docs/cli/389936/run-pipelines-cli because of error: Expected component `TabItem` to be defined: you likely forgot to import, pass, or provide it.
  
  --------------------------
  
  Error: Docusaurus server-side rendering could not render static page with path /docs/cli/465058/install because of error: Expected component `TabItem` to be defined: you likely forgot to import, pass, or provide it.
  
  --------------------------
  
  Error: Docusaurus server-side rendering could not render static page with path /docs/quickstart/729236/cli because of error: Expected component `TabItem` to be defined: you likely forgot to import, pass, or provide it.
  
  --------------------------
  
  Error: Docusaurus server-side rendering could not render static page with path /docs/zenith/user/install because of error: Expected component `TabItem` to be defined: you likely forgot to import, pass, or provide it.
      at /home/vikram/public/vikram.dagger/website/node_modules/@docusaurus/core/lib/webpack/utils.js:207:24
      at /home/vikram/public/vikram.dagger/website/node_modules/webpack/lib/MultiCompiler.js:554:14
      at processQueueWorker (/home/vikram/public/vikram.dagger/website/node_modules/webpack/lib/MultiCompiler.js:491:6)
      at process.processTicksAndRejections (node:internal/process/task_queues:77:11)
}

[INFO] Docusaurus version: 3.0.0
Node version: v18.16.0
```

`TabItem` is a component of the default Docusaurus theme, so it looks like there is a general issue with importing components.

**Comment by vikram-dagger** [comment]
Maybe a red herring but according to a message in the Docusaurus discord, this error occurs when using swizzled components due to incompatibilities. AFAIK we are using swizzled components in `src/theme`, so maybe this is the reason? WDYT @slumbering @crjm ?

Discord message: https://discord.com/channels/398180168688074762/1154771869094912090/1177362861463453766

image

**Comment by crjm** [comment]
> Maybe a red herring but according to a message in the Docusaurus discord, this error occurs when using swizzled components due to incompatibilities. AFAIK we are using swizzled components in src/theme, so maybe this is the reason? WDYT @slumbering @crjm ?

Yep, we are using swizzled components for sure. I'll look into it.

**Comment by vikram-dagger** [comment]
> Removes Docusaurus includes plugin in favour of native Docusaurus markdown import mechanism.
> Removes Docusaurus dotenv plugin as it does not seem to be required.
> Flattens directory structure for Docusaurus custom components as there are now fewer components and therefore additional categorization using sub-directories seems unnecessary (moves components from src/components/[atoms,molecules]/* to src/components and CSS files from src/css/[atoms,molecules]/* to src/css).
> Removes the custom theme files in `src/theme` and uses the default Docusaurus theme.

@slumbering @crjm I'm interested in your thoughts on above changes and any issues you see there.

**Comment by vikram-dagger** [comment]
> It seems like the [install CLI](https://github.com/dagger/dagger/blob/0bb93f09c41faefb2ddefe468a11e4df6b08078e/website/docs/current/partials/_install-cli.md) file is not importing the `<TabItem>` component, and it's relying on the parent imports (in this case `/docs/cli/389936/run-pipelines-cli`).
> 
> Apparently, the error being printed to the console is a bit misleading, because it throws the error on the parent files, instead of the partial. Just adding the imports at the top of the partial and removing them from the parents importing the partial is enough to make the error go away.
> 
> BTW, I would advise to move all the `.md` files that import components to the `.mdx` extension, as stated [here](https://docusaurus.io/docs/migration/v3#using-the-mdx-extension).

@crjm, thanks! This was super helpful in unblocking me. Please see the list of changes made and provide additional feedback as mentioned in https://github.com/dagger/dagger/pull/6164#issuecomment-1832493579

**Comment by crjm** [comment]
> Please see the list of changes made and provide additional feedback as mentioned in https://github.com/dagger/dagger/pull/6164#issuecomment-1832493579

I'm seeing that we'll need to change the base directory in Netlify from `/website` to `/docs`, otherwise builds won't get triggered. This should be the last step of the PR, as modifying this now in Netlify will block PRs that modify the `/website` directory. In this case, we'll need to wait until this is merged, and change the directory after merging. This won't cause down time, as the last built version is the one that will remain published.
The downside with this approach is that we won't be able to share previews for this migration.

An alternative can be keeping the `/website` directory, and add a build option in the `netlify.toml` that points to the `/docs` directory as the build path. This way, we should have previews available, and we can delete the `/website` directory after this gets merged.

cc @gerhard 

---

On another note, the quickstart is not rendering the loader while the iframe is loading. This leaves a white screen for a while until the snippet loads. I could help with bringing it back tomorrow, but don't want to block the PR.

**Review by slumbering — APPROVED** [review]

**Comment by vikram-dagger** [comment]
Once we merge and deploy this, the /0.1/* and /0.2/* URLs will produce 404s. We discussed two options:
- building a separate site from the 2.x branch; however afaik this will have the required content but it will not be served at /0.1 and /0.2 since this URL scheme was introduced with the 3.x branch
- building the content from this branch, which already has configuration for the /0.1 and /0.2 URL scheme

Also, an open question remains as to where the 0.1 and 0.2 docs archive will live ie. should we create `archive.docs.dagger.io/0.[1,2]/*` or `docs.dagger.io/archive/0.[1,2]/*` or similar?

Finally, regardless of which option we choose, URL redirections will need to be implemented for the migrated 0.1 and 0.2 docs.

**Comment by slumbering** [comment]
@vikram-dagger, I've updated the Netlify deployment directory (only for devel.docs...), the preview works fine now

**Comment by slumbering** [comment]
> Once we merge and deploy this, the /0.1/* and /0.2/* URLs will produce 404s. We discussed two options:
> 
> * building a separate site from the 2.x branch; however afaik this will have the required content but it will not be served at /0.1 and /0.2 since this URL scheme was introduced with the 3.x branch
> * building the content from this branch, which already has configuration for the /0.1 and /0.2 URL scheme
> 
> Also, an open question remains as to the URL where the 0.1 and 0.2 docs archive will live ie. should we create `archive.docs.dagger.io/0.[1,2]/*` or similar? We discussed that ideally this should be a separate site which is rebuilt only occasionally.
>
> Finally, regardless of which option we choose, URL redirections will need to be implemented for the migrated 0.1 and 0.2 docs.

Is there anything stopping us from using the Docusaurus versioning system? When I check their repository, they use it:
https://github.com/facebook/docusaurus/tree/main/website

Capture d’écran 2023-12-04 à 09 48 51
Capture d’écran 2023-12-04 à 09 49 07
Capture d’écran 2023-12-04 à 09 49 01

**Comment by vikram-dagger** [comment]
> Also, an open question remains as to the URL where the 0.1 and 0.2 docs archive will live ie. should we create `archive.docs.dagger.io/0.[1,2]/*` or similar? We discussed that ideally this should be a separate site which is rebuilt only occasionally.

Final decision is to have a separate subdomain (maybe archive.docs.dagger.io or similar) with URL paths like /0.1, /0.2 etc which will host the static docs sub-sites for those versions.

> Finally, regardless of which option we choose, URL redirections will need to be implemented for the migrated 0.1 and 0.2 docs.

We realized this can be done quite easily using Netlify redirections, so we will try this.

File: sdk/python/src/dagger/client/gen.py
Comment (slumbering) [reply sdk/python/src/dagger/client/gen.py:1018]: I'm doing the same thing for each SDKs

**Comment by gerhard** [comment]
Just finished reviewing the preview & everything works as expected apart from:
image

As soon as we have fixed the failing checks, this is good to go 🚀

**Comment by vikram-dagger** [comment]
> Just finished reviewing the preview & everything works as expected apart from: 

I looked into why the TOC doesn't appear on the "Contributing" page in your screenshot, while it does on the current Docusaurus 2 site. 

The reason is that we are importing the [Dagger repository's CONTRIBUTING.md](https://github.com/dagger/dagger/blob/main/CONTRIBUTING.md) file as a partial, and Docusaurus 3 does not detect headings in imported partials. This is documented at https://docusaurus.io/docs/markdown-features/react#importing-markdown and there is an open issue for it at https://github.com/facebook/docusaurus/issues/3915. I suspect we have to live with the limitation until a fix arrives in Docusaurus itself.

**Comment by gerhard** [comment]
Great find! OK to do as a follow-up, most likely via a Docusaurus upgrade.

**Comment by vikram-dagger** [comment]
Note: the Netlify preview is expected to fail on this PR as the build output directory is now docs/build while Netlify is looking for it in website/build (old). Once this PR is merged, we will update the Netlify configuration for the new build output path.

**Comment by gerhard** [comment]
I am just starting to look at this.

Will test locally, understand that we will make the Netlify change after merging.

**Comment by gerhard** [comment]
Are the `Edit this page` links working as expected @vikram-dagger?

I see this one pointing to https://github.com/dagger/dagger/edit/main/website/current_docs/quickstart/index.mdx

image

**Comment by gerhard** [comment]
Slight variation in the code preview colour scheme (left is before this PR, right is after), but this is expected.  The new one looks better IMO!

image

**Comment by gerhard** [comment]
Another LGTM page:
image

**Comment by gerhard** [comment]
Also good!

image

**Comment by gerhard** [comment]
Another important page working as expected:

image

**Comment by gerhard** [comment]
Minor colour & wording differences for danger/warning sections, but this is OK:

image

**Comment by gerhard** [comment]
The FAQ sidebar is now working as expected in this PR (didn't look right before). All good, just pointing it out.

image

**Review by gerhard — APPROVED** [review]
What an epic PR! Thank you @vikram-dagger for seeing it through 🙌 💯 🚀

**Comment by vikram-dagger** [comment]
> Are the `Edit this page` links working as expected @vikram-dagger?
> 
> I see this one pointing to https://github.com/dagger/dagger/edit/main/website/current_docs/quickstart/index.mdx

Thanks for catching this, fixed now!

**Pull request merged**
